src/lib/define-dynamic-block.js

```diff
@@ -21,6 +22,46 @@ const defineDynamicBlock = (ScratchBlocks, categoryInfo, staticBlockInfo, extend
         if (staticBlockInfo.blockIconURI || categoryInfo.blockIconURI) {
             blockJson.extensions = ['scratch_extension'];
         }
+
+        // Handle custom context menu options
+        // TODO this should probably not live here in the future, or at least
```

Comment:
I'd say this should probably be a separate function but it's fine to be in this file, called from this location. In my mind the problem of re-registering context menus is basically the same as the existing problem we have with re-registering blocks. I think we might just have to accept the `delete ScratchBlocks.somewhere[blockType]` / `delete ScratchBlocks.somewhereElse[contextMenuName]` kind of approach. Would that be sufficient in the menu case, or is there more we need to do in order to "unregister" a context menu?

Reply:
I haven't tested this out, but I think we might be able to do something smarter with it (with a little bit of effort that I did not think was worth it for the time being)... 

Looking at the data category as an example from scratch blocks (https://github.com/LLK/scratch-blocks/blob/develop/blocks_vertical/data.js#L51), we can have a map of contextMenu options per block to do something similar to what we're doing for the `scratch_extension` ScratchBlocks extension above.

Reply:
Not sure if that comment was totally clear. I was thinking we'd do something similar to how we handle block menus where we define the shadow blocks for menus separately from the specific blocks specified by the extensions. These blocks are defined only once during `handleExtensionAdded` in the blocks container. We could probably similarly have a blockContextMenus section of the _extension info_ and register the mixins only once when the extension is loaded.

Reply:
@cwillisf, I just reorganized the code so that there is a separate helper function for the custom context menu setup. I've also changed the structure of define dynamic block a bit so that it calls the setup function before returning the block definition object (containing the `init`, `domToMutation` and `mutationToDom` functions). This means that the context menu registration attempt is happening way less often now (previously it was running every time you move the block on the workspace), but it's still not completely de-duplicated for the same reason as the blocks getting redefined issue.

Tracking in #TBD

Reply:
I forgot to mention, I recommend viewing these changes with the `hide whitespace changes` option checked because of the reorganized code.